lib/charms/layer/jenkins/api.py

```diff
@@ -165,21 +165,29 @@ def _make_client(self):
         user = creds.username()
         token = creds.token()
 
-        # TODO: also handle regenerated tokens
         if token is None:
-            client = jenkins.Jenkins(self.url, user, creds.password())
-            # If we're using Jenkins >= 2.129 we need to request a new token.
-            jenkins_version = self._packages.jenkins_version()
-            if LooseVersion(jenkins_version) >= LooseVersion('2.129'):
-                token = client.run_script(GET_NEW_TOKEN_SCRIPT.format(user)).strip()
-            else:
-                token = client.run_script(GET_LEGACY_TOKEN_SCRIPT.format(user)).strip()
-            creds.token(token)
+            creds.token(self._get_token(user, creds.password(), self._packages.jenkins_version()))
 
         client = jenkins.Jenkins(self.url, user, token)
-        client.get_whoami()
+        try:
+            client.get_whoami()
+        # Handling token regeneration when the current token is invalid.
+        # Otherwise, re-raise the exception.
+        except jenkins.JenkinsException as e:
+            if "401" in str(e):
+                creds.token(self._get_token(user, creds.password(), self._packages.jenkins_version()))
+            raise
```

Comment:
This "raise" is still going to execute even if the exception contains a 401. Is the what we want? The comment above makes it sound like it isn't.

Reply:
The comment was wrong indeed. I've changed it.

Reply:
The comment was wrong indeed. I've changed it.